### ISSUE TITLE
Add support for GetQueuedCompletionStatusEx()

### ISSUE THREAD

**Issue by TheStormN** [chriskohlhoff/asio#560]
`GetQueuedCompletionStatusEx()` function is available since Vista, so it should be pretty safe to rely on it. The primary advantage is that it can fetch multiple IOCP events at single call, which can greatly minimize syscalls for high performance servers, thus improving their throughput.

**Comment by mabrarov** [comment]
Here is an old discussion about the same in asio-users mailing list - [Using GetQueuedCompletionStatusEx instead of GetQueuedCompletionStatus](https://sourceforge.net/p/asio/mailman/asio-users/thread/000601cca2cd%24917374c0%24b45a5e40%24%40mail.ru/#msg28396716).

My understanding is that with `GetQueuedCompletionStatusEx()` we can fail to distribute events effectively b/w multiple threads executing `asio::io_context::run()` method of the same instance of `asio::io_context` class, i.e. the most efficient approach for Windows - "single io_context instance with multiple threads" - may lead to uneven loading of threads if completion handlers differ in terms of time to execute. This can also lead to degradation of performance.

At the same time, `GetQueuedCompletionStatusEx()` should lead to increased performance of "multiple io_context instances" approach, where there is a single thread per each instance of `asio::io_context` class executing `asio::io_context::run()` method.

Taking into account that:

1. Windows associate all threads calling `GetQueuedCompletionStatus` for the same instance of IOCP as a thread pool and uses LIFO for scheduling of these treads, i.e. (should) effectively schedule threads when using "single io_context instance with multiple threads" approach.
1. [`asio::io_context::io_context(int concurrency_hint)`](https://www.boost.org/doc/libs/1_74_0/doc/html/boost_asio/reference/io_context/io_context/overload2.html) constructor can be used to hint Asio what threading approach is used.

I believe that we can use both `GetQueuedCompletionStatus()` and `GetQueuedCompletionStatusEx()` functions and choose b/w them inside `asio::io_context` (inside `asio::detail::win_iocp_io_context` class) depending on `concurrency_hint`.

**Comment by TheStormN** [comment]
Yep, I think most common case for high perf servers is multiple io_contexts with 1 thread per context. If that optimization can be made as part of `concurrency_hint` that would be best. There are already defines for other platforms which optimize io_context for single-threaded usage.

@chriskohlhoff Any thoughts? If I craft a PR about this, would you even consider it?

**Comment by TheStormN** [comment]
Btw @mabrarov I've just payed a bit closer look at the docs of `GetQueuedCompletionStatusEx()` and it seems that we don't need to bother with `GetQueuedCompletionStatus()` at all. Basically the `Ex` version will check how many threads are currently waiting on the completion port and it will balance the work items between all of them accordingly.

**Comment by mabrarov** [comment]
> Basically the `Ex` version will check how many threads are currently waiting on the completion port and it will balance the work items between all of them accordingly.

Could you please share a link to documentation stating that?

**Comment by TheStormN** [comment]
https://docs.microsoft.com/en-us/windows/win32/fileio/getqueuedcompletionstatusex-func

`Server applications may have several threads calling the GetQueuedCompletionStatusEx function for the same completion port. As I/O operations complete, they are queued to this port in first-in-first-out order. If a thread is actively waiting on this call, one or more queued requests complete the call for that thread only.`

**Comment by mabrarov** [comment]
When opening pull request, please take care to support operating systems which Asio supports. Accrording to [Using Boost.Asio](https://www.boost.org/doc/libs/1_74_0/doc/html/boost_asio/using.html) which references:

> Windows XP, Windows Server 2003, and earlier versions of Windows. When running on Windows Vista, Windows Server 2008

we have to support at least Windows XP.

**Comment by TheStormN** [comment]
@mabrarov I always thought this is just outdated documentation, I would be really surprised if ASIO can be build with Visual Studio which supports targeting Windows 95. And if I remember correctly VS 2005 was the last version to support 95, 98 and Me.

**Comment by mabrarov** [comment]
I agree (I feel the same), but according to [Using Boost.Asio](https://www.boost.org/doc/libs/1_74_0/doc/html/boost_asio/using.html) Visual C++ 9.0 (Visual Studio 2008) is still supported, which means that Asio (potentially) supports Windows 95. I would like @chriskohlhoff to comment on that (the reason I raised this topic here). I'm not sure if Boost 1.74 still supports pre-C++11 compilers even. Let me check Windows XP and Visual Studio 2008 (I have VM for that).

**Comment by TheStormN** [comment]
But Visual C++ 9.0 Does not support 95/95/Me targeting. That's according to MS.

**Comment by mabrarov** [comment]
Hmm.. you are right. [Visual Studio 2008 Breaking Changes](https://docs.microsoft.com/en-us/cpp/porting/visual-cpp-change-history-2003-2015?view=vs-2019#visual-studio-2008-breaking-changes):

> The Windows 95, Windows 98, Windows ME, and Windows NT platforms are no longer supported. These operating systems have been removed from the list of targeted platforms.

So it means that we have to support at Windows XP and later, doesn't it?

**Comment by TheStormN** [comment]
True, as I mentioned when I created the issue, the Ex version is available since Windows Vista.
